client/src/legacy/LeftAndMain.EditForm.js

```diff
@@ -110,28 +110,89 @@ $.entwine('ss', function($){
     },
     'from .cms-tabset': {
       onafterredrawtabs: function () {
-        // Show validation errors if necessary
-        if(this.hasClass('validationerror')) {
-          // Ensure the first validation error is visible
-          var tabError = this.find('.message.validation, .message.required').first().closest('.tab');
-          $('.cms-container').clearCurrentTabState(); // clear state to avoid override later on
-
-          // Attempt #1: Look for nearest .ss-tabset (usually nested deeper underneath a .cms-tabset).
-          var $tabSet = tabError.closest('.ss-tabset');
-
-          // Attempt #2: Next level in tab-ception, try to select the tab within this higher level .cms-tabset if possible
-          if (!$tabSet.length) {
-            $tabSet = tabError.closest('.cms-tabset');
-          }
 
-          if ($tabSet.length) {
-            $tabSet.tabs('option', 'active', tabError.index('.tab'));
-          } else if (!this.getValidationErrorShown()) {
-            // Ensure that this error message popup won't be added more than once
-            this.setValidationErrorShown(true);
-            errorMessage(ss.i18n._t('Admin.VALIDATIONERROR', 'Validation Error'));
+        // This function will:
+        // - Add an invalid icon on each tab that contains form fields than failed validation
+        // - Set an alert message in the edit form error banner
+        const iconClass = 'font-icon-attention-1 tab-attention';
+        const iconTitle = ss.i18n._t(
+          'VALIDATION_ERRORS_IN_TAB',
+          'This tab contains validation errors.'
+        );
+        const alertMessageText = ss.i18n._t(
+          'Admin.VALIDATION_ERRORS_ON_PAGE',
+          'There are validation errors on this page, please fix them before saving or publishing.'
+        );
+        const toastNotificationMessage = ss.i18n._t(
+          'Admin.VALIDATIONERROR',
+          'Validation Error'
+        );
+        const $editFormErrorBanner = $("#Form_EditForm_error");
+
+        // Remove any existing invalid tab icons
+        this.find(`.${iconClass.split(' ').join('.')}`).remove();
+
+        // Check if there are any form validation errors
+        // Note - When DataObject::validate() fails validation, it does not set validationerror,
+        // instead it will populate #Form_EditForm_error in the returned HTML server-side
+        let validationErrorExists = false;
+        if (this.hasClass('validationerror')) {
+          // Validation errors from DataObject::getCMSValidator()
+          // $editFormErrorBanner.html() will be blank at this point
+          validationErrorExists = true;
+        } else {
+          // Validation errors from DataObject::validate();
+          // 'validationerror' class will not be set at this point
+          if ($editFormErrorBanner.html() !== '') {
+            validationErrorExists = true;
           }
         }
```

Comment:
It could, though I also want to leave instructions to future developers of how the 3x different validation errors manifest themselves on the frontend.  Having it verbose like this makes it easy to add inline comments